### ISSUE TITLE
Thoughts on the default mappings

### ISSUE THREAD

**Issue by ashfinal** [mg979/vim-visual-multi#86]
First of all, thanks for your awesome plugin! Works much better than multiple-cursors, IMO. In the meanwhile, one thing that gets in the way of integrating visual-multi into my workflow is the default key mappings. They are so complicated, actually it took me one whole day to figure out what they are for and how to customize them. Anyway, now everything clears up seemingly.

From what I see, the problem with the default mappings is **we mixed the global mappings, or permanent mappings in this circumstance, with the buffer-related mappings when it concerned the *leader* key**, which confuses users very much. Don't get me wrong here, the **leader** key is good, as long as there is only one classification. **But the thing is we've gotten two here, that needs to be distinguished.** With that, people are clearly aware on which level they are now and which key mappings they can access to.

Also, we need to respect the setting of the **global mapleader**. The global mappings are those we need to access to everywhere everytime, that is the exact scenario where the mapleader kicks in. And fortunately, we can manage to limit the number of visual-multi related mappings to 3 or 5, which is perfect, minimizes the compact on users's existing configuration.

We still get use for `g:VM_leader`, mainly for buffer-related mappings. I would prefer `z` key, because in rare cases we need fold or spell functionality while using visual-multi plugin.

So, in a summary, we use `<Leader>key` for global manipulation, and `z` `key` for region/pattern related actions. I believe this helps clear up the confusion, reduce memory burden. And here is [my key mappings for visual-multi](https://macplay.github.io/listings/vimrc.before.html), if you're interested.

It's awkward to have 100 lines configuration before using this awesome plugin. If you like my proposal, I'll create a PR based on this, make it usable out of the box as best as I can. Thanks again for the plugin. :)

**Comment by mg979** [comment]
The thing with the mappings is a long-standing issue that, if one gets to understand how it all works, it's a non-issue, really. This is my configuration, for example:

```vim
let g:VM_mouse_mappings             = 1
let g:VM_leader                     = '\'
let g:VM_custom_remaps              = {'<c-p>': 'N'}
let g:VM_custom_noremaps            = {'))': '])', ')}': ']}'}
let g:VM_maps                       = {}
let g:VM_maps["Undo"]               = 'u'
let g:VM_maps["Redo"]               = '<C-r>'
let g:VM_maps["Select Cursor Down"] = '<M-C-Down>'
let g:VM_maps["Select Cursor Up"]   = '<M-C-Up>'
let g:VM_maps["Toggle Mappings"]    = '<Space>'
```

You can see I have `C-p` remapped as `N` so that I can use `C-n` and `C-p` to get next/previous.

I had a quick look to your mappings and my thoughts are:

* where did you take the mappings from, wiki or help documentation? I'd avoid the wiki atm.
* I don't want conflicts with default vim bindings for permanent mappings, so `z`  can't be the default VM leader. 
* if you use `z` a lot in mappings, why don't you just use `z` as `g:VM_leader`? That would reduce a lot the mappings to set.
* if you want to use `z` as leader for *buffer mappings* only, you can remap the permanent mappings, and use `z` as `g:VM_leader`. There are maybe a couple of permanent mappings that also use the VM leader, so you should remap those (and ideally *only those* ones).
* I see you set in your remaps the `<Leader>`, why don't you just set the real key?

If you think in terms of 2 leaders it all becomes more complicated, just think: VM leader is whatever key you want for most VM commands, and the other ones (especially the permanent mappings) you set them to whatever key you are more confortable with, without thinking in terms of `leader`.

Yes, it's very awkward to have a configuration that long, it shouldn't be that way, ideally you would only remap the leader, and maybe some selected mappings. Also, don't remap stuff that you don't use, or you'll get fed up. Stick with the most obvious stuff at first and if you feel the need to add something specific later, you can do it later.

**Comment by ashfinal** [comment]
I get the mappings mostly from the help docs and source code, luckily for me, they are all in the same file [all.vim](https://github.com/mg979/vim-visual-multi/blob/master/autoload/vm/maps/all.vim).

Conflicts with default vim bindings is what every plugin tries to avoid. But they still remapped `S/s` in surround.vim, didn't they? :) But that's not our scenario here. By entering Cursor/Extend mode, we are basically creating a gloal mode. And we've got `s` (select operator) and `m` (find operator) there, a `z` may slightly conflict with default vim bindings, but helps us a lot on the other hand.

Yes, I was to use `z` as `g:VM_leader`，until I found some global mappings also used it as prefix, so I commented it out. BTW, I also have this in my `.vimrc`:

``` vim
" With a map leader it's possible to do extra key combinations
let mapleader = "\<Space>"
```

> The issue in this is, maybe, that the VM leader is also used in some permanent mappings.

Yes, that exectly is my point. This mixture confuses people, and helps little to mental memory. Sure, I can work this around, half down my mappings configuration, which will still hold up to, say, 30 lines at best cases?

The fact is I think the default settings are important, and we have the chance to do things right from the beginning.

**Comment by mg979** [comment]
I think what you need is something like this:

```vim
    let g:VM_maps = {}
    let g:VM_leader                        = 'z'
    let g:VM_maps["Get Operator"]          = '\gs'
    let g:VM_maps["Reselect Last"]         = '\gS'
    let g:VM_maps["Add Cursor At Pos"]     = '\\'
    let g:VM_maps["Start Regex Search"]    = '\/'
    let g:VM_maps["Select All"]            = '\A'
    let g:VM_maps["Visual Regex"]          = '\/'
    let g:VM_maps["Visual All"]            = '\A'
    let g:VM_maps["Visual Add"]            = '\a'
    let g:VM_maps["Visual Find"]           = '\f'
    let g:VM_maps["Visual Cursors"]        = '\c'
```

Anyway I think I'll do this: you'll be able to define the `g:VM_leader` as a single key (like `\`) or as a list (like `['\', 'z']`, in the second case the first key would be used for permanent mappings, the second key for buffer mappings. It would be backwards compatible and would solve this kind of issues, sounds ok?

**Comment by mg979** [comment]
I implemented as a dictionary in the test branch, if you want to try it, see `:help g:VM_leader-dict`
You can do for example

    let g:VM_leader = {'default': '/', 'visual': '/', 'buffer': 'z'}

So that the prefixes are different for permanent/visual and buffer mappings.

**Comment by ashfinal** [comment]
Defining `g:VM_leader` as a dictionary is a good idea. It is backwards compatible, and allows me to cut off lots of visual-multi configuration.
``` vim

let g:VM_leader = {'default': '<Leader>', 'visual': '<Leader>', 'buffer': 'z'}

let g:VM_maps                           = {}
let g:VM_maps["Get Operator"]           = '<Leader>a'
let g:VM_maps["Add Cursor At Pos"]      = '<Leader><Space>'
let g:VM_maps["Reselect Last"]          = '<Leader>z'

let g:VM_maps["Visual Cursors"]         = '<Leader><Space>'
```

[Here is the full version](https://macplay.github.io/listings/vimrc.before.html).

I also find a little glitch there. Since `Visual Subtract/Reduce` only have meanings after entering VM mode, they are likely buffer mappings. Same case goes with `Erase Regions`. I've create a PR based this.

**Comment by mg979** [comment]
Just for the record, many buffer mappings actually used the `z` key in the past, before I opted for the `VM_leader`, so this has been a good idea after all. Thanks, please reopen if you want to add something.

**Comment by ashfinal** [comment]
Yeah, actually I found the only two keys left without intervening in the matter of editing and jumping, is `m` and `z`. Unless there is another important operator appearing in the future, using `z` as prefix is a good choice for now, helps cleaning up the whole key mappings.

**Comment by masaeedu** [comment]
@mg979 Is https://github.com/mg979/vim-visual-multi/commit/23d3a3910469a6f81de5b4bdb485d092c2eb7e78 a breaking change? I ran a `:PlugUpdate` today and my keybindings had stopped working. After doing a `git bisect` it seems like this commit is where it stopped working specifically. If it is intentionally a breaking change, would it be possible to add this sort of thing to a `CHANGELOG` file or similar?

Here's what I had configured:

```
let g:VM_maps = {}
let g:VM_maps["Add Cursor Down"]='<S-Down>'
let g:VM_maps["Add Cursor Up"]='<S-Up>'
let g:VM_maps["Move Right"]='<M-S-Right>'
let g:VM_maps["Move Left"]='<M-S-Left>'
```
